presto-hive/src/main/java/io/prestosql/plugin/hive/ReaderProjectionsAdapter.java

```diff
@@ -0,0 +1,245 @@
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.prestosql.plugin.hive.ReaderProjectionsAdapter.ChannelMapping.createChannelMapping;
+import static io.prestosql.spi.block.ColumnarRow.toColumnarRow;
+import static java.util.Objects.requireNonNull;
+
+public class ReaderProjectionsAdapter
+{
+    private final List<ChannelMapping> outputToInputMapping;
+    private final List<Type> outputTypes;
+    private final List<Type> inputTypes;
+
+    public ReaderProjectionsAdapter(List<HiveColumnHandle> expectedHiveColumns, ReaderProjections readerProjections)
+    {
+        requireNonNull(expectedHiveColumns, "expectedHiveColumns is null");
+        requireNonNull(readerProjections, "readerProjections is null");
+
+        ImmutableList.Builder<ChannelMapping> mappingBuilder = ImmutableList.builder();
+
+        for (int i = 0; i < expectedHiveColumns.size(); i++) {
+            HiveColumnHandle projectedColumnHandle = readerProjections.readerColumnForHiveColumnAt(i);
+            int inputChannel = readerProjections.readerColumnPositionForHiveColumnAt(i);
+            ChannelMapping mapping = createChannelMapping(expectedHiveColumns.get(i), projectedColumnHandle, inputChannel);
+            mappingBuilder.add(mapping);
+        }
+
+        outputToInputMapping = mappingBuilder.build();
+
+        outputTypes = expectedHiveColumns.stream()
+                .map(column -> column.getType())
+                .collect(toImmutableList());
+
+        inputTypes = readerProjections.getReaderColumns().stream()
+                .map(column -> column.getType())
+                .collect(toImmutableList());
+    }
+
+    public Page adaptPage(Page input)
+    {
+        if (input == null) {
+            return null;
+        }
+
+        Block[] blocks = new Block[outputToInputMapping.size()];
+
+        // Prepare adaptaions to extract dereferences
+        for (int i = 0; i < outputToInputMapping.size(); i++) {
+            ChannelMapping mapping = outputToInputMapping.get(i);
+
+            Block inputBlock = input.getBlock(mapping.getInputChannelIndex());
+            blocks[i] = createAdaptedLazyBlock(inputBlock, mapping.getDereferenceSequence(), outputTypes.get(i));
+        }
+
+        return new Page(input.getPositionCount(), blocks);
+    }
+
+    private static Block createAdaptedLazyBlock(Block inputBlock, List<Integer> dereferenceSequence, Type type)
+    {
+        if (dereferenceSequence.size() == 0) {
+            return inputBlock;
+        }
+
+        if (inputBlock == null) {
+            return null;
+        }
+
+        return new LazyBlock(inputBlock.getPositionCount(), new DereferenceBlockLoader(inputBlock, dereferenceSequence, type));
+    }
+
+    private static class DereferenceBlockLoader
+            implements LazyBlockLoader
+    {
+        boolean loaded;
```

Comment:
`private`

Also, move the final fields before the mutable ones (no strict convention about this, but we generally tend to organize them that way).